### PR TITLE
i#1569 AArch64: Print instruction if encoding fails.

### PR DESCRIPTION
**Pull request by fhahn**
As the parser for AArch64 is not complete yet dumping instructions that could
not be encoded makes it easier to diagnose encoding problems.

### CODE REVIEW

**Comment by fhahn** [comment]
`252 - code_api|api.static_signal (Failed) ` caused the Travis-ci failure: https://travis-ci.org/DynamoRIO/dynamorio/jobs/216403717

Xref #2112

**Comment by derekbruening** [comment]
Make sure any issue hit on Travis or Appveyor has the Hotlist-Travis label.  At some point we'll have a Travis Fixit and try to get all-hand-on-board for a burst of flaky test fixing, using that label to find all such problems.

**Comment by derekbruening** [comment]
I added the label.

File: core/arch/aarch64/codec.c [hunk 1]
```diff
@@ -46,7 +46,6 @@
 
 #include "codec.h"
 
-#define ENCFAIL (uint)0 /* a value that is not a valid instruction */
```
Comment (derekbruening) [inline core/arch/aarch64/codec.c]: Isn't ENCFAIL used in codec.c below?  Wouldn't this make it no longer compile?

File: core/arch/aarch64/encode.c [hunk 2]
```diff
@@ -160,6 +160,11 @@ instr_encode_arch(dcontext_t *dcontext, instr_t *instr, byte *copy_pc, byte *fin
     CLIENT_ASSERT(instr_operands_valid(instr), "instr_encode error: operands invalid");
 
     *(uint *)copy_pc = encode_common(final_pc, instr);
+    if (*(uint *)copy_pc == ENCFAIL) {
+        IF_DEBUG(instr_disassemble(dcontext, instr, STDERR));
```
Comment (derekbruening) [inline core/arch/aarch64/encode.c]: Everything printed should go through the checks for stderr_mask.  How about instr_disassemble_to_buffer and then print via SYSLOG_INTERNAL_ERROR which has the proper checks and also tees the print to the log file.

File: core/arch/aarch64/codec.c (hunk 1 above)
Comment (fhahn) [inline core/arch/aarch64/codec.c]: I moved the define to codec.h, which is included by codec.c

File: core/arch/aarch64/encode.c (hunk 2 above)
Comment (fhahn) [inline core/arch/aarch64/encode.c]: Will do

File: core/arch/aarch64/codec.c (hunk 1 above)
Comment (derekbruening) [inline core/arch/aarch64/codec.c]: Hmm, I misread it, thought it was moved to encode.c.

File: core/arch/aarch64/encode.c (hunk 2 above)
Comment (fhahn) [inline core/arch/aarch64/encode.c]: Done in latest commit. I've also added a short message using SYSLOG_INTERNAL_ERROR, it should be much better now than just using instr_disassemble :)

File: core/arch/aarch64/encode.c [hunk 3]
```diff
@@ -161,8 +161,14 @@ instr_encode_arch(dcontext_t *dcontext, instr_t *instr, byte *copy_pc, byte *fin
 
     *(uint *)copy_pc = encode_common(final_pc, instr);
     if (*(uint *)copy_pc == ENCFAIL) {
-        IF_DEBUG(instr_disassemble(dcontext, instr, STDERR));
         /* We were unable to encode this instruction. */
+        IF_DEBUG({
```
Comment (derekbruening) [inline core/arch/aarch64/encode.c]: DODEBUG is what's generally used but the difference is minor.

**Comment by derekbruening** [comment]
I think my recent change to static_signal may have made it flaky on Travis.

**Comment by derekbruening** [comment]
It looks like the bottom of the commit message was not fully edited on the squash-and-merge: please remember to edit the message, as github just squashes the different commit messages together.

**Comment by derekbruening** [comment]
There's also weird whitespace:

```
Date:   Thu Mar 30 16:54:49 2017 +0100

    i#1569 AArch64: Print instruction if encoding fails. (#2315)
    
    
    
    As the parser for AArch64 is not complete yet dumping instructions that could
    not be encoded makes it easier to diagnose encoding problems.
    
    * Use SYSLOG_INTERNAL_ERROR to dump disassembled instruction.
```

**Pull request merged**